### PR TITLE
 feat: enhance quality calculation (WPB-24323)

### PR DESCRIPTION
**Pull request by sifa-wire**
WPB-24323: Iterate on network quality index

Quality index is calculated [wrt latency, packet loss and jitter.](https://wearezeta.atlassian.net/wiki/spaces/PAD/pages/2520252449/2026-01-29+RFC+Call+Details+Network+Quality+Interface)
All stats are 3 level quantized wrt min and max thresholds.
Upstream metrics are weighted more than downstrem ones (0.7 vs 0.3)
Over quality index is calculated with weighted average;

NQI = 0.35 * loss + 0.35 * latency + 0.3 * jitter

### CODE REVIEW

File: src/wcall/wcall.c [hunk 1]
```diff
@@ -1621,6 +1622,42 @@ static void destructor(void *arg)
 	info("wcall(%p): dtor -- done\n", wcall);
 }
 
+static int normalize_to_levels(int num, int low_threshold, int high_threshold) {
+	if (num < low_threshold) {
+		return 1;
```
Comment (z-dule) [inline src/wcall/wcall.c]: Please, no hard-coded values, don't we have constants for quality levels? WCALL_QUALITY_NORMAL etc?

File: src/wcall/wcall.c [hunk 2]
```diff
@@ -1621,6 +1622,42 @@ static void destructor(void *arg)
...
+static int normalize_quality(const struct stats_report* stats) {
+	// Stats are 3 step normalized wrt corresponding thresholds
+	const int rtt = normalize_to_levels(stats->rtt, RTT_LOW, RTT_HIGH);
+	const int jitter_tx = normalize_to_levels((stats->jitter.audio.tx + stats->jitter.video.tx) / 2, JITTER_LOW, JITTER_HIGH);
+	const int jitter_rx = normalize_to_levels((stats->jitter.audio.rx + stats->jitter.video.rx) / 2, JITTER_LOW, JITTER_HIGH);
+	const int packet_loss_tx = normalize_to_levels(stats->packets.lost.tx, PACKET_LOSS_LOW, PACKET_LOSS_HIGH);
+	const int packet_loss_rx = normalize_to_levels(stats->packets.lost.rx, PACKET_LOSS_LOW, PACKET_LOSS_HIGH);
+
+	// provide higher importance to upstream stats
+	float jitter = 0.7 * jitter_tx + 0.3 * jitter_rx;
```
Comment (z-dule) [inline src/wcall/wcall.c]: Please, no hard-coded values, use constans instead, will make any tweaking much easier...

File: src/wcall/wcall.c [hunk 3]
```diff
@@ -1621,6 +1622,42 @@ static void destructor(void *arg)
...
+	float packet_loss = 0.7 * packet_loss_tx + 0.3 * packet_loss_rx;
+
+	// provide packet loss and jitter a bit more importance than latency
+	return round(0.35 * jitter + 0.35 * packet_loss + 0.3 * rtt);
```
Comment (z-dule) [inline src/wcall/wcall.c]: Again, please, no hard-coded values...

**Review by z-dule — CHANGES_REQUESTED** [review]

File: src/wcall/wcall.c (hunk 1 above)
Comment (sifa-wire) [reply src/wcall/wcall.c]: handled with 504aeed121882bf2771dc402f55d289e6da2d29a

File: src/wcall/wcall.c (hunk 2 above)
Comment (sifa-wire) [reply src/wcall/wcall.c]: handled with 504aeed121882bf2771dc402f55d289e6da2d29a

File: src/wcall/wcall.c (hunk 3 above)
Comment (sifa-wire) [reply src/wcall/wcall.c]: handled with 504aeed121882bf2771dc402f55d289e6da2d29a

**Review by z-dule — APPROVED** [review]
